Feature/toasts

tobias-wiese opened the pull request:
# Description

- Closes #495 (for feature implementation)

## Reason for this PR
- Toasts improve the user experience

## Changes in this PR
- add "vue-toastification"
- implement global toasts that are usable from within the setup method *and* ts helper files

### Dependency update
- add ```"vue-toastification": "^2.0.0-beta.5"```

## Type of change (remove all that don't apply)
- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Manually tested all use cases this could alter

**Test Configuration**:
- OS: Windows Mac Linux
- Browser: Firefox Chrome Safari Chromium

- Frontend: (remove all that don't apply)
  - [x] Development build
- Backend: (remove all that don't apply)
  - [x] No backend needed to test this
  - [x] deployed development build

# Checklist: (remove all that don't apply)

- [x] I have performed a self-review of my own code
- [x] My changes generate no new linting warnings or console warnings
- [x] I have updated the CHANGELOG.md to include any changes made in this PR (add WIP to the top, if there is none already)

## Notes:
- the timeouts of the toast did not work on my local setup for some reason. I haven't found a solution for that yet, but I think it has sth to do with the CSS

bastihav reviewed (changes requested):
- there are a lot of "warning" toasts, which are used to display errors.. why not use the error toast?
- We should talk about (next meeting) how many toasts we really want to use. In my opinion, these are too many. 
- toasts not closing as an open issue (with a pending PR) https://github.com/Maronato/vue-toastification/issues/137 - we should wait for the fix
- Personally, I would change "showAPI400Toast(), showAPI404Toast()" etc to 
  - a enumeration (there is probably already a HTTP status code enum we could use)
  - a generic function showToast(code: HTTPStatusCodeEnumStuff), which then uses a switch case, but this is a minor opinionated change.
- HTTP Bad Request is not an adequate user feedback, when we fucked up. "Something went wrong on our side. Please consider to report this bug" + a clickable report button similar to the one on our about page (toasts can carry arbitrary vue components)
- Same for HTTP not Found, "Could not find user/course/whatever) might be more suitable, but not as generic, so we would need to pass some more information
- Things like toast.warning("Error: could not delete profile picture.");
  - This does not help a user in solving the problem. I think toasts like this should only be issued inside the response handlers, as this is either
    - a network problem
    - a 404
    - a 403
    - ...? (refer to swagger API, which always lists all error codes, you can ignore authentication)
  - all of which should be handled in a generic way, not on a per API call basis. In the current case, this would just spam the user with toasts: "Network error, bla bla" "Could not delete profile picture"


Just a few thoughts, let's talk about this again.

tobias-wiese commented:
> 
> 
>     * there are a lot of "warning" toasts, which are used to display errors.. why not use the error toast?

Warning toasts are removed
> 
>     * We should talk about (next meeting) how many toasts we really want to use. In my opinion, these are too many.
> 
number of toasts is reduced significantly.
>     * toasts not closing as an open issue (with a pending PR) [Maronato/vue-toastification#137](https://github.com/Maronato/vue-toastification/issues/137) - we should wait for the fix
> 
fixed.
>     * Personally, I would change "showAPI400Toast(), showAPI404Toast()" etc to
>       
>       * a enumeration (there is probably already a HTTP status code enum we could use)
>       * a generic function showToast(code: HTTPStatusCodeEnumStuff), which then uses a switch case, but this is a minor opinionated change.
I implemented a generic function and extended generec response handlers by a ```dataType: string``` that describes its usage for showing 404 error messages. The generic ```showAPIToast(code: string, msg?: string)``` function supports HTTP codes 400 and 404 atm.
We decided to not use an HTTP code enum as it would just map eg. ```"404": 404```
> 
>     * HTTP Bad Request is not an adequate user feedback, when we fucked up. "Something went wrong on our side. Please consider to report this bug" + a clickable report button similar to the one on our about page (toasts can carry arbitrary vue components)
Implemented with a custom component with the functions from the about page
> 
>     * Same for HTTP not Found, "Could not find user/course/whatever) might be more suitable, but not as generic, so we would need to pass some more information
see above
> 
>     * Things like toast.warning("Error: could not delete profile picture.");
>       
>       * This does not help a user in solving the problem. I think toasts like this should only be issued inside the response handlers, as this is either
>         
>         * a network problem
>         * a 404
>         * a 403
>         * ...? (refer to swagger API, which always lists all error codes, you can ignore authentication)
>       * all of which should be handled in a generic way, not on a per API call basis. In the current case, this would just spam the user with toasts: "Network error, bla bla" "Could not delete profile picture"

I removed the error toasts that were not need because of existing UI error feedback.


here screenshot of the existing toasts.
![grafik](https://user-images.githubusercontent.com/63233799/97293255-b64cce00-184c-11eb-9de5-6acd51e40d20.png)

bastihav commented on src/components/common/ReportBug.vue:
```diff
@@ -0,0 +1,63 @@
+<template>
+    <div class="flex">
+        <label>Something went wrong on our side. Please consider to report this bug</label>
```
```suggestion
        <label>Something went wrong on our side. Please consider reporting this bug.</label>
```

bastihav commented on src/use/helpers/ImmatriculationResponseHandler.ts:
```diff
@@ -1,13 +1,50 @@
+                return response.returnValue;
+            }
+            case 403: {
+                alert("You do not have the neccessary user rights for this action!");
```
this should be changed

tobias-wiese replied:
this does not belong on this branch. It is just marked as added lines, because I merged two files. Please consider to open an issue for this

bastihav commented on src/use/helpers/Toasts.ts:
```diff
@@ -0,0 +1,32 @@
+    toast.error("Error: '" + subject + "' not updated.");
+}
+
+export function showAPIToast(code: string, msg?: string) {
```
why is code a string, when it is only ever used for numbers?

tobias-wiese replied:
I changed it to number.

bastihav reviewed (changes requested):
- You use e.g. toast.success("Password updated."); (multiple locations where you directly invoke toasts), while you have defined showUpdateSuccessToast(subject: string) to abstract this
- In those abstraction methods, you directly inject a string into the sentence, so when adding a new feature, we would always need to remember how the sentence is structured, would it make sense to use an enumeration here? So I only need to worry about "showUpdateSuccessToast(Enum.COURSE)"? This would also allow us to change the sentence later and just adjust the enumeration, instead of changing the parameter in all components.
- Maybe we should also add abstraction for toast.success("Course '" + course.value.courseName + "' created."); etc, I think if we keep all sentences that are shown to the user in one place will make it much easier to change things consistently in all components (also with regards to translation, but thats post PG talk)
- I don't think I like the parameter for response handlers, but I don't know any better way to do it. Maybe we should also think about using an enumeration here, same reasoning as before.

tobias-wiese commented:
> 
> 
>     * You use e.g. toast.success("Password updated."); (multiple locations where you directly invoke toasts), while you have defined showUpdateSuccessToast(subject: string) to abstract this
> 
>     * In those abstraction methods, you directly inject a string into the sentence, so when adding a new feature, we would always need to remember how the sentence is structured, would it make sense to use an enumeration here? So I only need to worry about "showUpdateSuccessToast(Enum.COURSE)"? This would also allow us to change the sentence later and just adjust the enumeration, instead of changing the parameter in all components.
> 
>     * Maybe we should also add abstraction for toast.success("Course '" + course.value.courseName + "' created."); etc, I think if we keep all sentences that are shown to the user in one place will make it much easier to change things consistently in all components (also with regards to translation, but thats post PG talk)
This could be a nice idea to keep in mind. But with enums, we will not be able to show the name of the created in the toast. Using enums, we could just display eg "Course created". Additionally, I do not think, that we will have multiple appearances of such success messages.

tobias-wiese commented:
@bastihav @theovier This branch is now ready to be reviewed and merged. The usage of the custom component for an api code 500 is replaced by a placeholder for now until #696 is fixed.

theovier reviewed (approved):
I think we can increase the default show-time of modals just by a little bit.

bastihav commented on CHANGELOG.md:
```diff
@@ -24,6 +24,9 @@
 
 # [v.0.11.0](https://github.com/upb-uc4/ui-web/compare/v0.10.0...v0.11.0) (2020-10-29)
 
+## Feature
+- add toasts for user interaction [#613](https://github.com/upb-uc4/ui-web/pull/613)
+
```
needs to be moved (can't do it via suggestion)
